### ISSUE TITLE
Extended buffering options

### ISSUE THREAD

**Issue by edsiper** [fluent/fluent-bit#2377]
Community is asking for extended options to control the buffering strategies. Opening this ticket to open the discussion and design the extended solution.

**Comment by JeffLuoo** [comment]
Move the discussion from closed issue #2365 to here.

Mentioned By @erain: 
What we are really working on is the offline buffer support for output plugins.

Below are our understanding and the feature we want:

The current use case for buffering is from input to output: if a message has been processed by a output plugin, it will be taken out of the buffer.
The output plugin may drop messages if its destination is unreachable.
Do you have any suggestion or guidance on the offline buffering for the output plugins (e.g. as a general fluent bit feature or as a feature in each of the output plugin)?

**Comment by edsiper** [comment]
@JeffLuoo 

Fluent Bit already provides a secondary buffering mechanism based in the file system. It can be enabled per-ingestor plugin, so every record  and chunk registered it keeps a copy in the file system until is fully delivered to the destination:

- https://docs.fluentbit.io/manual/administration/buffering-and-storage

Note that output plugins never drops a message, it only reports FLB_OK (success), FLB_RETRY (ask for a retry), FLB_ERROR (unrecoverable error).

Most of the time plugins asks for a FLB_RETRY, then the engine decides if it should schedule or not a retry based in the user configuration for that specific destination. 

Talking about the options proposed:

-  ```retry_timeout```: what would be the real use-case ?, when a retry is scheduled the "time" for that retry can be anything in terms of seconds, that is generated by a [backoff full jitter](https://github.com/fluent/fluent-bit/blob/master/src/flb_scheduler.c#L235) algorithm. That one exists because we don't want to have all the retries at the same time. Implementing a retry timeout can generate other bad side-effects.

- ```total_limit_size```: this sounds really good and should be something implemented under the ```storage.X``` property of the output plugins. when the output plugin asks for a retry, the engine might check for the oldest chunk to be dropped due to the restriction set by configuration. 

The first step would be to define the ideal storage.X properties to control the buffered data per plugin destination.

**Comment by JeffLuoo** [comment]
Proposal was created #2408

**Comment by HackToHell** [comment]
@edsiper I have one question regarding how Mem_Buf_Limit works, piggy backing on this question as I think it might be relevant, will, the engine not call the pause callback on the input plugin if the output plugin keeps on failing? How will the storage_limit interact with  backpressure.

**Comment by erain** [comment]
@HackToHell to my understanding the engine will keep re-trying following a [truncated exponential backoff algorithm](https://cloud.google.com/storage/docs/exponential-backoff). 

Current the engine will keep the oldest chunk once `storage_limit` is reached. We have enhance that in our proposal by some new configuration parameters.

**Comment by erain** [comment]
I think this is implemented as part of offline buffering feature and we can close it now.

**Comment by majoyz** [comment]
`retry_max_interval/retry_timeout` might be importent in unreliable networking environment.
It seems the chunks waiting to retry can't realize the output server(like influxdb) is reachable or not, if it wait too long, it might miss the opportunity like this:
```
influxDB: not ok     ok              not ok                  
          ___________****************______________
chunk:    |retry failed, waiting...       |retry failed,waiting...
```
The chunks might have more opportunity to retry succeed by limit the max waiting time, like 5min.

**Comment by lololozhkin** [comment]
Hi, is this idea is still alive? From my point of view it's very important functionality, because exponential algorithm grows too fast. After few retries timeout becomes about 1000s (more than 10 minutes) and it is too big interval for unreliable network. retry_max_interval option from fluentd is great solution for these problems.

Are you going to do this in next update, or should I better use fluentd?

**Comment by JeffLuoo** [comment]
@majoyz @lololozhkin Please take a look of #4099 to see if this is what you need.

**Comment by lololozhkin** [comment]
> @majoyz @lololozhkin Please take a look of #4099 to see if this is what you need.

Yes, thank you! It was exactly what I wanted to configure, but I wasn't able to find this parameter in documentation.
